### PR TITLE
Serialization and Deserialization for RawMessage

### PR DESCRIPTION
**Pull request by ivankelly**
Serializes the message ID and passes through the payload and metadata
of the raw message.

### CODE REVIEW

**Comment by ivankelly** [comment]
@merlimat @sijie

**Comment by ivankelly** [comment]
Implements issue #965

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawMessageImpl.java [hunk 1]
```diff
@@ -53,4 +60,47 @@ public ByteBuf getHeadersAndPayload() {
...
+        int headerSize = 4 /* IdSize */ + idSize + 4 /* PayloadAndMetadataSize */;
+
+        ByteBuf headers = PooledByteBufAllocator.DEFAULT.buffer(headerSize);
+        headers.writeInt(idSize);
+        try {
+            ByteBufCodedOutputStream outStream = ByteBufCodedOutputStream.get(headers);
+            id.writeTo(outStream);
+            outStream.recycle();
+        } catch (IOException e) {
+            // This is in-memory serialization, should not fail
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawMessageImpl.java]: nit: log a message here?

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawMessageImpl.java [hunk 2]
```diff
@@ -53,4 +60,47 @@ public ByteBuf getHeadersAndPayload() {
...
+            MessageIdData id = builder.mergeFrom(stream, null).build();
+            buffer.writerIndex(writerIndex);
+            builder.recycle();
+
+            int payloadAndMetadataSize = buffer.readInt();
+            ByteBuf metadataAndPayload = buffer.slice(buffer.readerIndex(), payloadAndMetadataSize);
+
+            return new RawMessageImpl(id, metadataAndPayload);
+        } catch (IOException e) {
+            // This is in-memory deserialization, should not fail
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawMessageImpl.java]: nit: log a message here?

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawMessageImpl.java (hunk 1 above)
Comment (ivankelly) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawMessageImpl.java]: done

File: pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawMessageImpl.java (hunk 2 above)
Comment (ivankelly) [reply pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawMessageImpl.java]: done

**Comment by ivankelly** [comment]
@sijie @merlimat ping

**Pull request merged**